include/fault.h

```diff
@@ -0,0 +1,102 @@
+#ifndef FAULT_H
+#define FAULT_H
+
+#include "ultra64.h"
+#include "libc/stdarg.h"
+#include "libc/stdint.h"
+#include "io/controller.h"
+
+// These are the same as the 3-bit ansi color codes
+#define FAULT_COLOR_BLACK      0
+#define FAULT_COLOR_RED        1
+#define FAULT_COLOR_GREEN      2
+#define FAULT_COLOR_YELLOW     3
+#define FAULT_COLOR_BLUE       4
+#define FAULT_COLOR_MAGENTA    5
+#define FAULT_COLOR_CYAN       6
+#define FAULT_COLOR_WHITE      7
+// Additional color codes
+#define FAULT_COLOR_DARK_GRAY  8
+#define FAULT_COLOR_LIGHT_GRAY 9
+
+#define FAULT_COLOR_STRINGIFY(s) #s
+#define FAULT_COLOR_EXPAND_AND_STRINGIFY(s) FAULT_COLOR_STRINGIFY(s)
+
+#define FAULT_ESC '\x1A'
+#define FAULT_COLOR(n) "\x1A" FAULT_COLOR_EXPAND_AND_STRINGIFY(FAULT_COLOR_ ## n)
+
+
+#define NORMAL_RDRAM_END 0x80400000
+// Address at the end of normal RDRAM after which is room for a screen buffer
+#define FAULT_FB_ADDRESS (NORMAL_RDRAM_END - SCREEN_HEIGHT * SCREEN_WIDTH * sizeof(u16))
+
+
+typedef struct FaultClient {
+    /* 0x0 */ struct FaultClient* next;
+    /* 0x4 */ void (*callback)(void*, void*);
+    /* 0x8 */ void* arg0;
+    /* 0xC */ void* arg1;
+} FaultClient; // size = 0x10
+
+typedef struct FaultAddrConvClient {
+    /* 0x0 */ struct FaultAddrConvClient* next;
+    /* 0x4 */ uintptr_t (*callback)(uintptr_t, void*);
+    /* 0x8 */ void* arg;
+} FaultAddrConvClient; // size = 0xC
+
+typedef void(*FaultPadCallback)(Input* input);
+
+// Initialization
+
+void Fault_Init(void);
+
+// Fatal Errors
+
+void Fault_AddHungupAndCrashImpl(const char* exp1, const char* exp2);
+void Fault_AddHungupAndCrash(const char* filename, u32 line);
+
+// Client Registration
+
+void Fault_AddClient(FaultClient* client, void* callback, void* arg0, void* arg1);
+void Fault_RemoveClient(FaultClient* client);
+void Fault_AddAddrConvClient(FaultAddrConvClient* client, void* callback, void* arg);
+void Fault_RemoveAddrConvClient(FaultAddrConvClient* client);
+
+// For use in Fault Client callbacks
+
+void Fault_WaitForInput(void);
+void Fault_FillScreenBlack(void);
+void Fault_SetFrameBuffer(void* fb, u16 w, u16 h);
+
+void FaultDrawer_SetForeColor(u16 color);
+void FaultDrawer_SetBackColor(u16 color);
+void FaultDrawer_SetFontColor(u16 color);
+void FaultDrawer_SetCharPad(s8 padW, s8 padH);
+void FaultDrawer_SetCursor(s32 x, s32 y);
+s32 FaultDrawer_VPrintf(const char* fmt, va_list ap);
+s32 FaultDrawer_Printf(const char* fmt, ...);
+void FaultDrawer_DrawText(s32 x, s32 y, const char* fmt, ...);
+
+
+typedef struct FaultMgr {
+    /* 0x000 */ OSThread thread;
+    /* 0x1B0 */ u8 stack[0x600]; // Seems leftover from an earlier version. The thread actually uses a stack of this size at 0x8009BE60
```

Comment:
Out of curiosity if this seems like a stack variable, would it use the `STACK` macro?

Reply:
I was thinking in using the `STACK` macro, but I would need #1193 to be merged first. If this PR goes in first then I could change this in that PR